File: solutions/javascript/list-ops/1/list-ops.js
export class List {
  #values = [];

  constructor(data) {
    if (data) this.#values = [...data];
  }

  get values() {
    return this.#values;
  }

  append(second) {
    return new List([...this.#values, ...second.#values]);
  }

  // Mutates lhs in order to enable recursion
  static concat(lhs, rhs) {
    for (let elem of rhs.values) {
      if (elem instanceof List == false) {
        lhs.#values.push(elem);
      } else List.concat(lhs, elem);
    }
    return this;
  }

  concat(second) {
    let result = new List([...this.values]);
    List.concat(result, second);
    return result;
  }

  filter(fn) {
    return new List(this.values.filter(fn));
  }

  map(fn) {
    return new List(this.values.map(fn));
  }

  length() {
    return this.values.length;
  }

  foldl(fn, acc_init) {
    return this.values.reduce(fn, acc_init);
  }

  foldr(fn, acc_init) {
    const temp = [...this.values];
    temp.reverse();
    return temp.reduce(fn, acc_init);
  }

  reverse() {
    const temp = [...this.values];
    temp.reverse();
    return new List(temp);
  }
}

File: solutions/python/list-ops/1/list_ops.py
from collections.abc import Iterable
from typing import Callable, TypeVar, List

T = TypeVar("T")
U = TypeVar("U")


def append(list1: list[T], list2: list[T]) -> list[T]:
    return [*list1, *list2]


def concat(lists: list[list[T]]) -> list[T]:
    result = []
    for list_ in lists:
        result.extend(list_)
    return result


def filter(function: Callable[[T], bool], list: list[T]) -> list[T]:
    return [el for el in list if function(el)]


def length(list: list) -> int:
    return len(list)


def map(function: Callable[[T], U], list: list[T]) -> list[U]:
    return [function(el) for el in list]


def foldl(function: Callable[[T, U], T], list: list[U], initial: T) -> T:
    result = initial
    for el in list:
        result = function(result, el)
    return result


def foldr(function: Callable[[T, U], T], list: list[U], initial: T) -> T:
    result = initial
    for el in reversed(list):
        result = function(result, el)
    return result


def reverse(list: list[T]) -> list[T]:
    return list[::-1]
